src/persist/src/indexed/mod.rs

```diff
@@ -1562,6 +1634,122 @@ mod tests {
+    /// run.
+    fn trace_compaction_unexpected_response() -> Result<(), Error> {
+        let unreliable = UnreliableHandle::default();
+        let (mut i, maintainer, id) = trace_compaction_setup(unreliable)?;
+
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        let request = reqs[0].clone();
+
+        // Send an error response. Now there are no more requests in flight.
+        let error_response =
+            MaintenanceRes::CompactTrace((id, Err(Error::from("test compaction error"))));
+        i.handle_maintenance_res(error_response);
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // Handle receiving an unexpected (valid) MaintenanceRes when none is in flight.
+        let response = get_maintenance_response(&maintainer, request);
+        i.handle_maintenance_res(response);
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        Ok(())
+    }
+
+    /// This test checks that Indexed correctly handles receiving an error MaintenanceRes
+    /// even when no requests are in flight for that stream. This test is expected to
+    /// panic when run.
+    fn trace_compaction_unexpected_error_response() -> Result<(), Error> {
+        let unreliable = UnreliableHandle::default();
+        let (mut i, _, id) = trace_compaction_setup(unreliable)?;
+
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+
+        // Send an error response. Now there are no more requests in flight.
+        let error_response =
+            MaintenanceRes::CompactTrace((id, Err(Error::from("test compaction error"))));
+        i.handle_maintenance_res(error_response.clone());
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // Handle receiving an unexpected error MaintenanceRes when no request is in flight.
+        i.handle_maintenance_res(error_response);
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        Ok(())
+    }
+
+    #[test]
+    fn trace_compaction_expected_errors() -> Result<(), Error> {
+        fn check_error(
+            e: Option<Box<dyn Any + Send + 'static>>,
+            expected: &str,
+        ) -> Result<(), Error> {
+            let e = match e {
+                Some(e) => e,
+                None => return Err(Error::from("unexpected success while checking error")),
+            };
+
+            let error = match e.downcast_ref::<String>() {
+                Some(error) => error.clone(),
+                None => return Err(Error::from("unable to check error")),
+            };
+
+            assert!(
+                error.contains(expected),
+                "{} not found in {}",
+                expected,
+                error
+            );
+            Ok(())
+        }
+        let result = std::panic::catch_unwind(|| {
+            let _ = trace_compaction_unexpected_response();
+        });
+
+        check_error(result.err(), "received trace compaction response for a stream that did not have a in flight request: Id(0). Ignoring")?;
```

Comment:
edit the whitespace to make it clearer that we're checking the error for the previous result